core/src/main/java/org/apache/iceberg/mapping/MappingUtil.java

```diff
@@ -91,7 +92,14 @@ public MappedFields mapping(NameMapping mapping, MappedFields result) {
 
     @Override
     public MappedFields fields(MappedFields fields, List<MappedField> fieldResults) {
-      return MappedFields.of(fieldResults);
+      ImmutableMap.Builder<String, Integer> builder = ImmutableMap.builder();
+      fieldResults.stream()
+          .map(MappedField::id).filter(Objects::nonNull)
```

Comment:
No, what's happening here is that we have a set of changes to the schema by ID. We want the changes that apply just to this struct in the schema, so that we don't reassign names that changed in other structs.

To get the changes that apply to this mapping level, we use ids from the existing mappings. A a mapping may not have an id, indicating that we have not yet mapped an incoming name, but that the name exists in an associated name-based schema. Filtering to non-null ids gets rid of those fields. Then we look up the ids for this mapping in the list of changes, filter out unchanged fields where the result of the lookup is null, and apply the results.

Reply:
I guess you could argue that `changes.get(null)` should be null, but it seems strange to me to look up null in a map. You're right that we could probably remove it. Do you think we should?